cmds/webboot/webboot.go

```diff
@@ -0,0 +1,245 @@
+type DownloadOption struct {
+	isoName string
+	url     string
+}
+
+// IsDefault is for mark whether this iso is a default choise.
+func (u ISO) IsDefault() bool {
+	return u.isDefault
+}
+
+func (u ISO) String() string {
+	return fmt.Sprintf("%+v\n", u)
+}
+
+// Label is the string this iso displays in the menu page.
+func (u ISO) Label() string {
+	return u.name
+}
+
+// IsDefault is for mark whether this entry is a default choise. we assume download option always be a default option.
+func (u DownloadOption) IsDefault() bool {
+	return true
+}
+
+func (u DownloadOption) String() string {
+	return fmt.Sprintf("%+v\n", u)
+}
+
+// Label is the string this entry displays in the menu page. download option will be displayed as "Download" in the menu page.
+func (u DownloadOption) Label() string {
+	return "Download"
+}
+
+// getKernel is to find the kernel of a ISO. if the path of ISO is given, just check if the kernel is exist.
+func getKernel(u *ISO) error {
+
+	log.Println("try mount iso...")
+
+	isoName := u.name
+
+	diskFile, err := ioutil.TempDir("", "/mnt-iso")
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer os.RemoveAll(diskFile)
+
+	mountPath := filepath.Join(diskFile, isoName)
+	if mp, err := mount.Mount(u.path, mountPath, "iso9660", "", mount.ReadOnly); err != nil {
+		log.Printf("TryMount %s = %v, want nil", u.path, err)
+	} else {
+		log.Printf("mounted disk file is %s\n", mountPath)
+		// if kernel is given, check its validation
+		if u.kernel != "" {
+			walkfunc := func(path string, info os.FileInfo, err error) error {
+				if path == filepath.Join(mountPath, u.kernel) {
+					log.Printf("Kernel at %s is found\n", u.kernel)
+				}
+				return nil
+			}
+
+			log.Println("\nfinding kernel...")
+			filepath.Walk(diskFile, walkfunc)
+			log.Printf("Kernel path is %s\n", u.kernel)
+		}
+
+		log.Println("\ntry unmount iso...")
+		if err := mp.Unmount(0); err != nil {
+			log.Printf("Unmount(%q) = %v, want nil", mountPath, err)
+		}
+	}
+
+	log.Println("Done")
+	return nil
+}
+
+// isoMenu is to  build the iso menu page and return the iso path and kernel path of which user choose.
+func isoMenu(isos []ISO) (string, string) {
+
+	var entries []menus.Entry
+	for _, iso := range isos {
+		entries = append(entries, iso)
+	}
+	var downloadOption DownloadOption
+	entries = append(entries, downloadOption)
+
+	index, err := menus.DisplayMenu("ISO Menu", "Choose an iso you want to boot (hit enter to choose the default - 1 is the default option) >", 0, entries)
+
+	// if index is exceed the length of entries or less than 0 means something wrong happened
+	if err != nil || index < 0 || index >= len(entries) {
+		if err != nil {
+			log.Println(err)
+		}
+		return "", ""
+	}
+	// if index is the same as the length of entries - 1, it means the chosen option is download. Elsewise the chosen option is a cached iso.
+	if index == len(entries)-1 {
```

Comment:
I commented this somewhere else too, but if DisplayMenu returned the entry chosen instead of just the index, this logic could be cleaner

Reply:
I hope I can return the entry, but entry do not contain enough information. And because different option need different information (like a iso option need the path of iso and the download option need the input name of iso) i can not make a entry work for all options. that is entry only contains option's label and whether it is a default option, but I need more information in most of situations, that's why I return index instead of entry.

Reply:
Maybe the way to deal with this is to modify the Entry interface to also have a Do() method like you used to have before, so that you can call Do() directly from the menu function. So you'd probably need to do something like

```
func (u ISO) Do() error {
	err := getKernel(u)
        // Add logic for actually booting the entry
}
```

Reply:
I thought about this but the problem is if I run getKernel() in Do(). I don't know how to inform my main func what is getKernel() return. I may need the path to kernel return by getKernel() in next step to replace the kernel in main. And I also think about let Do() return what getKernel() return, but the problem is, in the Entry struct I have declare the format of Do(), but different kind of option (like iso kind and download kind) will want different kind of information so I can not define the format of Do() clearly.

Reply:
So I think the idea here is that Do() should take care of all the steps that happen after an ISO has been selected. This includes finding the kernel and initrd (in our case, this will be done using FindDevices from u-root), and kexec. The only reason why you'd need to go back to your main function is if there was an error, which can be dealt with by returning the error. So this would probably look like

```
type Entry interface {
   Label() String
   IsDefault() bool
  Do() error
}

func (u ISO) Do() error {
	err := getKernel(u)
        // Add logic for actually booting the entry
}

func (u ISO) Do() error {
	err := getKernel(u)
        // Add logic for actually booting the entry
}
```

Does that work?

Reply:
Thank you! I will try this way!